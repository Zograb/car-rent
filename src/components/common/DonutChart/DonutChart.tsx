import type { DonutChartDataItem } from '@/types';
import type { CSSProp } from 'styled-components';
import { type FC, useCallback, useMemo } from 'react';
import {
  PieChart,
  Pie,
  Cell,
  ResponsiveContainer,
} from 'recharts';
import tw from 'twin.macro';

// Utils
import { getRandomColor } from '@/utils';

export interface DonutChartProps {
  // container styles for chart and legend
  containerStyle?: CSSProp;

  // class for chart container
  className?: string;

  // show/hide chart legend
  showLegend?: boolean;

  // determine which side to render chart legend
  legendPosition?: 'bottom' | 'right';

  // chart data to render cells
  data: DonutChartDataItem[];
}

export const DonutChart: FC<DonutChartProps> = ({
  containerStyle = tw``,
  className = '',
  legendPosition = 'bottom',
  data,
  showLegend,
}) => {
  // Generate random color if not provided
  const colors: string[] = useMemo(() => (
    data.map(((d: DonutChartDataItem) => d.color || getRandomColor()))
  ), [data]);

  const totalValue: number = useMemo(() => {
    let value = 0;
    for (let i = 0; i < data.length; i++) {
      value += data[i].value;
    }
    return value;
  }, [data]);

  const getPercentOfTotal = useCallback((value: number): string => {
    const percent = Math.round(value * 100 / totalValue);
    return `${percent}%`;
  }, [totalValue]);

  return (
    <div
      css={[
        legendPosition === 'right' && tw`flex justify-start items-center`,
        containerStyle,
      ]}
    >
      <div
        css={[
          tw`w-44 h-44`,
          legendPosition === 'bottom' && tw`mx-auto`,
        ]}
        className={className}
      >
        <ResponsiveContainer width="100%" height="100%">
          <PieChart width={400} height={400}>
            <Pie
              dataKey="value"
              isAnimationActive={true}
              data={data}
              fill="#8884d8"
              innerRadius={45}
              stroke="none"
              data-testid="chart"
            >
              {colors.map((color, index) => <Cell key={`cell-${index}`} fill={color} />)}
            </Pie>
          </PieChart>
        </ResponsiveContainer>
      </div>
      {showLegend && (
        <div
          css={[
            legendPosition === 'bottom' && tw`mt-4`,
            legendPosition === 'right' && tw`ml-4 grow`,
          ]}
          data-testid="donut-chart-legend"
        >
          {data.map((d, index) => (
            <div tw="flex items-center mb-2" key={d.name}>
              <div
                tw="rounded-full w-4 h-4"
                style={{ backgroundColor: d.color || colors[index] }}
              />
              <div tw="grow ml-3 capitalize">{d.name}</div>
              <div>{getPercentOfTotal(d.value)}</div>
            </div>
          ))}
        </div>
      )}
    </div>
  );
}
