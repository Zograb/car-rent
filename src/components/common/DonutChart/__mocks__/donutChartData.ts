const donutChartData = [
  {
    name: 'Part 1',
    value: 360,
    color: '#52C93F',
  },
  {
    name: 'Part 2',
    value: 230,
    color: '#006AFF',
  },
  {
    name: 'Part 3',
    value: 150,
    color: '#FF2727',
  },
  {
    name: 'Part 4',
    value: 230,
    color: '#FF2727',
  },
  {
    name: 'Part 5',
    value: 150,
    color: '#FFFF00',
  },
];

export default donutChartData;
