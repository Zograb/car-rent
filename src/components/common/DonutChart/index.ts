export { type DonutChartProps, default as DonutChart } from './DonutChart';
