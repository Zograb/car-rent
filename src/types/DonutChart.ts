export interface DonutChartDataItem {
  name: string;
  value: number;
  color?: string;
}
