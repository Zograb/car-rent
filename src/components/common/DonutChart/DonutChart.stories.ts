import type { Meta, StoryObj } from '@storybook/react';
import tw from 'twin.macro';

import DonutChart from './DonutChart';
import donutChartData from './__mocks__/donutChartData';

const meta: Meta<typeof DonutChart> = {
  title: 'Donut Chart',
  component: DonutChart,
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof DonutChart>;

export const Basic: Story = {
  args: {
    data: donutChartData,
  },
};

export const WithLegend: Story = {
  args: {
    data: donutChartData,
    showLegend: true,
    containerStyle: tw`w-60 mx-auto`,
  },
}

export const WithLegendOnRight: Story = {
  args: {
    data: donutChartData,
    showLegend: true,
    containerStyle: tw`mx-auto w-96`,
    legendPosition: 'right',
  },
}
