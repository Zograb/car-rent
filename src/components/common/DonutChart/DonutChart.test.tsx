import type { ReactElement } from 'react';
import { render, screen } from '@testing-library/react';
import recharts from 'recharts';

// Components
import DonutChart from './DonutChart';

// Mocks
import donutChartData from './__mocks__/donutChartData';
import {
  mockResizeObserverBefore,
  mockResizeObserverAfter,
} from '@/__mocks__';

// Set responsive container width and height to fixed value
// Otherwise in test envirinment it will not render chart at all
jest.mock('recharts', () => {
  const OriginalModule: typeof recharts = jest.requireActual('recharts');

  return {
    ...OriginalModule,
    ResponsiveContainer: ({ children }: { height: number, children: ReactElement }) => (
      <div style={{ width: 500, height: 500 }}>
        {children}
      </div>
    ),
  };
});

describe('DonutChart', () => {
  beforeEach(mockResizeObserverBefore);

  afterEach(mockResizeObserverAfter);

  it('should match snapshot', async () => {
    const { container } = render(<DonutChart data={donutChartData} />);
    expect(container).toMatchSnapshot();
  });

  it('should render chart', async () => {
    const { container } = render(<DonutChart data={donutChartData} />);
    const chart = container.querySelector('svg');
    expect(chart).toBeInTheDocument();
  });

  it('should render chart with legend', async () => {
    render(<DonutChart data={donutChartData} showLegend />);
    const chartLegend = screen.getByTestId('donut-chart-legend');
    expect(chartLegend).toBeInTheDocument();
  });
});

