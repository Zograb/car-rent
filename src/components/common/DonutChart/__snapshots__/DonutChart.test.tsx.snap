// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DonutChart should match snapshot 1`] = `
.c0 {
  height: 11rem;
  width: 11rem;
  margin-left: auto;
  margin-right: auto;
}

<div>
  <div
    class=""
  >
    <div
      class="c0"
    >
      <div
        style="width: 500px; height: 500px;"
      >
        <div
          class="recharts-wrapper"
          role="region"
          style="position: relative; cursor: default; width: 400px; height: 400px;"
        >
          <svg
            class="recharts-surface"
            cx="50%"
            cy="50%"
            height="400"
            viewBox="0 0 400 400"
            width="400"
          >
            <title />
            <desc />
            <defs>
              <clippath
                id="recharts1-clip"
              >
                <rect
                  height="390"
                  width="390"
                  x="5"
                  y="5"
                />
              </clippath>
            </defs>
            <g
              class="recharts-layer recharts-pie"
              tabindex="0"
            >
              <g
                class="recharts-layer"
              >
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
                <g
                  class="recharts-layer recharts-pie-sector"
                  tabindex="-1"
                />
              </g>
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
